client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -656,13 +678,22 @@ void reconnect() {
                               state.failConnection(e1);
                               throw Exceptions.sneakyThrow(e1);
                           }
+                          // A timeout is added to the future before the call, and it triggers a TimeoutException.
+                          // A late timeout if fine it will just cause a spurious connection close.
+                          // A late success may be a problem because it causes retransmits of the wrong messages.
+                          // In theory the server should guard against this, but that's not ideal to depend on for client correctness.
```

Comment:
This relates to the to 540 above. With this change we should be able to update the code there to fail the connection. Please make that change similar to the reverted PR and have the comments here and there cross reference one other.